File: zig/src/help.zig
const std = @import("std");

const HELP_TEXT =
    \\ccc — call coding CLIs
    \\
    \\Usage:
    \\  ccc [runner] [+thinking] [:provider:model] [@alias] "<Prompt>"
    \\  ccc --help
    \\  ccc -h
    \\
    \\Slots (in order):
    \\  runner        Select which coding CLI to use (default: oc)
    \\                opencode (oc), claude (cc), kimi (k), codex (rc), crush (cr)
    \\  +thinking     Set thinking level: +0 (off) through +4 (max)
    \\  :provider:model  Override provider and model
    \\  @alias        Use a named preset from config
    \\
    \\Examples:
    \\  ccc "Fix the failing tests"
    \\  ccc oc "Refactor auth module"
    \\  ccc cc +2 :anthropic:claude-sonnet-4-20250514 "Add tests"
    \\  ccc k +4 "Debug the parser"
    \\  ccc codex "Write a unit test"
    \\
    \\Config:
    \\  ~/.config/ccc/config.toml  — default runner, aliases, abbreviations
    \\
;

const RunnerEntry = struct {
    name: []const u8,
    alias: []const u8,
};

const CANONICAL_RUNNERS = [_]RunnerEntry{
    .{ .name = "opencode", .alias = "oc" },
    .{ .name = "claude", .alias = "cc" },
    .{ .name = "kimi", .alias = "k" },
    .{ .name = "codex", .alias = "rc" },
    .{ .name = "crush", .alias = "cr" },
};

fn getVersion(allocator: std.mem.Allocator, binary: []const u8) []const u8 {
    const argv = [_][]const u8{ binary, "--version" };
    var child = std.process.Child.init(&argv, allocator);
    child.stdin_behavior = .Ignore;
    child.stdout_behavior = .Pipe;
    child.stderr_behavior = .Ignore;

    child.spawn() catch return "";
    defer _ = child.wait() catch {};

    var stdout_buf: std.ArrayList(u8) = .empty;
    defer stdout_buf.deinit(allocator);

    child.stdout.?.reader().readAllArrayList(&stdout_buf, 4096) catch return "";

    const output = stdout_buf.items;
    if (output.len == 0) return "";

    if (std.mem.indexOfScalar(u8, output, '\n')) |nl| {
        return output[0..nl];
    }
    return output;
}

fn hasOnPath(allocator: std.mem.Allocator, name: []const u8) bool {
    var child = std.process.Child.init(
        &[_][]const u8{ "which", name },
        allocator,
    );
    child.stdin_behavior = .Ignore;
    child.stdout_behavior = .Ignore;
    child.stderr_behavior = .Ignore;

    child.spawn() catch return false;
    const term = child.wait() catch return false;
    return term.Exited == 0;
}

fn runnerChecklist(allocator: std.mem.Allocator, file: std.fs.File) !void {
    var writer_buf: [4096]u8 = undefined;
    var w = std.fs.File.Writer.init(file, &writer_buf);
    var io_writer = w.initInterface(&writer_buf);

    io_writer.writeAll("Runners:\n") catch return;

    for (CANONICAL_RUNNERS) |entry| {
        const on_path = hasOnPath(allocator, entry.name);
        if (on_path) {
            const version = getVersion(allocator, entry.name);
            if (version.len > 0) {
                io_writer.print("  [+] {s:<10} ({s})  {s}\n", .{ entry.name, entry.name, version }) catch return;
            } else {
                io_writer.print("  [+] {s:<10} ({s})  found\n", .{ entry.name, entry.name }) catch return;
            }
        } else {
            io_writer.print("  [-] {s:<10} ({s})  not found\n", .{ entry.name, entry.name }) catch return;
        }
    }
}

pub fn printHelp(allocator: std.mem.Allocator) !void {
    const stdout_file = std.fs.File.stdout();
    try stdout_file.writeAll(HELP_TEXT);
    try stdout_file.writeAll("\n");
    runnerChecklist(allocator, stdout_file) catch {};
}

pub fn printUsage(allocator: std.mem.Allocator) !void {
    const stderr_file = std.fs.File.stderr();
    try stderr_file.writeAll("usage: ccc [runner] [+thinking] [:provider:model] [@alias] \"<Prompt>\"\n");
    runnerChecklist(allocator, stderr_file) catch {};
}

File: cpp/src/help.cpp
#include <ccc/help.hpp>
#include <ccc/parser.hpp>

#include <array>
#include <cstdio>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>
#include <vector>

static const char* const kHelpText =
    "ccc \xe2\x80\x94 call coding CLIs\n"
    "\n"
    "Usage:\n"
    "  ccc [runner] [+thinking] [:provider:model] [@alias] \"<Prompt>\"\n"
    "  ccc --help\n"
    "  ccc -h\n"
    "\n"
    "Slots (in order):\n"
    "  runner        Select which coding CLI to use (default: oc)\n"
    "                opencode (oc), claude (cc), kimi (k), codex (rc), crush (cr)\n"
    "  +thinking     Set thinking level: +0 (off) through +4 (max)\n"
    "  :provider:model  Override provider and model\n"
    "  @alias        Use a named preset from config\n"
    "\n"
    "Examples:\n"
    "  ccc \"Fix the failing tests\"\n"
    "  ccc oc \"Refactor auth module\"\n"
    "  ccc cc +2 :anthropic:claude-sonnet-4-20250514 \"Add tests\"\n"
    "  ccc k +4 \"Debug the parser\"\n"
    "  ccc codex \"Write a unit test\"\n"
    "\n"
    "Config:\n"
    "  ~/.config/ccc/config.toml  \xe2\x80\x94 default runner, aliases, abbreviations\n"
    "\n";

const char* const HELP_TEXT = kHelpText;

static std::string getVersion(const std::string& binary) {
    std::string cmd = binary + " --version 2>/dev/null";
    FILE* pipe = popen(cmd.c_str(), "r");
    if (!pipe) return "";
    std::array<char, 256> buf;
    std::string result;
    while (fgets(buf.data(), static_cast<int>(buf.size()), pipe) != nullptr) {
        result += buf.data();
        auto nl = result.find('\n');
        if (nl != std::string::npos) {
            result.resize(nl);
            break;
        }
    }
    int status = pclose(pipe);
    if (status != 0 || result.empty()) return "";
    return result;
}

std::string runnerChecklist() {
    static const std::vector<std::pair<std::string, std::string>> runners = {
        {"opencode", "oc"},
        {"claude", "cc"},
        {"kimi", "k"},
        {"codex", "rc"},
        {"crush", "cr"},
    };

    const auto& registry = getRunnerRegistry();

    std::ostringstream oss;
    oss << "Runners:\n";
    for (const auto& [name, alias] : runners) {
        std::string binary = name;
        auto it = registry.find(name);
        if (it != registry.end()) {
            binary = it->second.binary;
        }
        if (auto fp = popen(("command -v " + binary + " 2>/dev/null").c_str(), "r")) {
            pclose(fp);
            std::string version = getVersion(binary);
            std::string tag = version.empty() ? "found" : version;
            oss << "  [+] " << name;
            for (size_t i = name.size(); i < 10; ++i) oss << ' ';
            oss << "(" << binary << ")  " << tag << "\n";
        } else {
            oss << "  [-] " << name;
            for (size_t i = name.size(); i < 10; ++i) oss << ' ';
            oss << "(" << binary << ")  not found\n";
        }
    }
    return oss.str();
}

void printHelp() {
    std::cout << HELP_TEXT << "\n" << runnerChecklist();
}

void printUsage() {
    std::cerr << "usage: ccc [runner] [+thinking] [:provider:model] [@alias] \"<Prompt>\"\n";
    std::cerr << runnerChecklist();
}
